Mage.Sets/src/mage/cards/t/TayamLuminousEnigma.java

```diff
@@ -0,0 +1,236 @@
+package mage.cards.t;
+
+import mage.MageInt;
+import mage.abilities.Ability;
+import mage.abilities.common.SimpleActivatedAbility;
+import mage.abilities.common.SimpleStaticAbility;
+import mage.abilities.costs.Cost;
+import mage.abilities.costs.common.RemoveCounterCost;
+import mage.abilities.costs.mana.GenericManaCost;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.ReplacementEffectImpl;
+import mage.abilities.effects.common.PutTopCardOfLibraryIntoGraveControllerEffect;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.choices.Choice;
+import mage.choices.ChoiceImpl;
+import mage.constants.*;
+import mage.counters.Counter;
+import mage.counters.CounterType;
+import mage.filter.FilterCard;
+import mage.filter.common.FilterControlledCreaturePermanent;
+import mage.filter.common.FilterPermanentCard;
+import mage.filter.predicate.mageobject.ConvertedManaCostPredicate;
+import mage.filter.predicate.permanent.CounterPredicate;
+import mage.game.Game;
+import mage.game.events.EntersTheBattlefieldEvent;
+import mage.game.events.GameEvent;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.TargetCard;
+import mage.target.TargetPermanent;
+import mage.target.common.TargetCardInYourGraveyard;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+/**
+ *
+ * @author htrajan
+ */
+public final class TayamLuminousEnigma extends CardImpl {
+
+    public TayamLuminousEnigma(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{1}{W}{B}{G}");
+
+        this.addSuperType(SuperType.LEGENDARY);
+        this.subtype.add(SubType.NIGHTMARE);
+        this.subtype.add(SubType.BEAST);
+        this.power = new MageInt(3);
+        this.toughness = new MageInt(3);
+
+        // Each other creature you control enters the battlefield with an additional vigilance counter on it.
+        this.addAbility(new SimpleStaticAbility(Zone.BATTLEFIELD, new TayamLuminousEnigmaReplacementEffect()));
+
+        // {3}, Remove three counters from among creatures you control: Put the top three cards of your library into your graveyard, then return a permanent card with converted mana cost 3 or less from your graveyard to the battlefield.
+        PutTopCardOfLibraryIntoGraveControllerEffect millEffect = new PutTopCardOfLibraryIntoGraveControllerEffect(3);
+        millEffect.concatBy(".");
+        Ability ability = new SimpleActivatedAbility(Zone.BATTLEFIELD, millEffect, new GenericManaCost(3));
+        ability.addCost(new TayamLuminousEnigmaCost());
+        TayamLuminousEnigmaEffect effect = new TayamLuminousEnigmaEffect();
+        effect.setText(", then return a permanent card with converted mana cost 3 or less from your graveyard to the battlefield");
+        ability.addEffect(effect);
```

Comment:
Did you check if the text comes out correctly?

Reply:
Yep!

![image](https://user-images.githubusercontent.com/4507238/79948865-80580880-8429-11ea-9439-45ca2902e912.png)